File: eval_harness/evaluator/skill_invocation/build_health/syntax_check.py
"""Syntax half of the build-health cascade: does every source file in the
authored app actually parse? Cheapest, most generic signal available --
catches broken code before spending anything on a real bundle/build.

Deliberately NOT a per-skill check (no skill_map.json entry, doesn't go
through the uptake_checks registry) -- this is app-wide, independent of
which skill(s) were expected, so it surfaces as its own top-level
"build_health" key in metrics.json rather than a tier under some skill.

Reuses uptake_checks' Node/Babel parser (node_parser.py) and its source-file
scanning (AppTree, same node_modules/scripts/lockfile exclusions as the
lexical checks) rather than duplicating either.
"""

from __future__ import annotations

from pathlib import Path

from ..uptake_checks.node_parser import parse_file_facts
from ..uptake_checks.registry import AppTree


def check_syntax(app_dir: Path) -> dict:
    app_tree = AppTree(app_dir)
    failed: list[dict] = []
    skipped_unavailable = 0
    for rel_path in app_tree.files:
        facts = parse_file_facts(app_dir / rel_path)
        if facts is None:
            skipped_unavailable += 1
            continue
        if "error" in facts:
            failed.append({"file": str(rel_path), "message": facts.get("message")})

    total = len(app_tree.files)
    checked = total - skipped_unavailable
    return {
        "total_files": total,
        "checked_files": checked,
        "skipped_unavailable": skipped_unavailable,
        "failed_files": failed,
        "ok": len(failed) == 0 if checked > 0 else None,
    }

File: eval_harness/evaluator/skill_invocation/build_health/syntax_check.ts
import { AppTree } from "../uptake_checks/registry.ts";
import { checkSourceSyntax } from "./node_parser.ts";

export type SyntaxCheckResult = {
  total_files: number;
  checked_files: number;
  skipped_unavailable: number;
  failed_files: Array<{ file: string; message: string }>;
  ok: boolean | null;
};

export async function checkSyntax(appDir: string): Promise<SyntaxCheckResult> {
  // AppTree performs directory discovery and source reads concurrently.
  // Promise.all preserves result order here, but Babel parsing remains
  // synchronous CPU work unless it is later moved into worker threads.
  const appTree = await AppTree.load(appDir);
  const checkedFiles = await Promise.all(
    [...appTree.files].map(async ([relativePath, source]) => ({
      relativePath,
      facts: checkSourceSyntax(source),
    })),
  );
  const failedFiles = checkedFiles.flatMap(({ relativePath, facts }) =>
    "error" in facts
      ? [{ file: relativePath, message: facts.message }]
      : [],
  );
  const total = appTree.files.size;
  return {
    total_files: total,
    checked_files: total,
    skipped_unavailable: 0,
    failed_files: failedFiles,
    ok: total > 0 ? failedFiles.length === 0 : null,
  };
}
